File: apps/android/feature/library/src/main/kotlin/app/storyarc/feature/library/SourcePresentation.kt
package app.storyarc.feature.library

import androidx.annotation.StringRes
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Dns
import androidx.compose.material.icons.filled.Folder
import androidx.compose.material.icons.filled.RssFeed
import androidx.compose.material.icons.filled.Storage
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import app.storyarc.core.designsystem.theme.StoryArcPalette
import app.storyarc.core.designsystem.tokens.StoryArcColor
import app.storyarc.core.model.SourceConnectionState
import app.storyarc.core.model.SourceKind

/**
 * Presentation for the domain's source types. Kept out of `core:model` so the
 * domain stays free of Compose — the same split iOS keeps between `StoryArcCore`
 * and `DesignSystem`.
 */
internal val SourceKind.icon: ImageVector
    get() = when (this) {
        SourceKind.LOCAL_FOLDER -> Icons.Filled.Folder
        SourceKind.NETWORK_SHARE -> Icons.Filled.Storage
        SourceKind.OPDS_CATALOG -> Icons.Filled.RssFeed
        SourceKind.KAVITA_SERVER -> Icons.Filled.Dns
    }

@get:StringRes
internal val SourceKind.titleRes: Int
    get() = when (this) {
        SourceKind.LOCAL_FOLDER -> R.string.source_kind_local_folder_title
        SourceKind.NETWORK_SHARE -> R.string.source_kind_network_share_title
        SourceKind.OPDS_CATALOG -> R.string.source_kind_opds_title
        SourceKind.KAVITA_SERVER -> R.string.source_kind_kavita_title
    }

@get:StringRes
internal val SourceKind.explanationRes: Int
    get() = when (this) {
        SourceKind.LOCAL_FOLDER -> R.string.source_kind_local_folder_explanation
        SourceKind.NETWORK_SHARE -> R.string.source_kind_network_share_explanation
        SourceKind.OPDS_CATALOG -> R.string.source_kind_opds_explanation
        SourceKind.KAVITA_SERVER -> R.string.source_kind_kavita_explanation
    }

@get:StringRes
internal val SourceConnectionState.statusRes: Int
    get() = when (this) {
        is SourceConnectionState.Connected -> R.string.source_state_connected
        is SourceConnectionState.Connecting -> R.string.source_state_connecting
        is SourceConnectionState.Unreachable -> R.string.source_state_unreachable
        is SourceConnectionState.Unauthorized -> R.string.source_state_unauthorized
    }

/**
 * Only `Unauthorized` is red. An unreachable source is grey, because `sources`
 * treats offline as a normal state rather than a failure.
 */
internal fun SourceConnectionState.indicatorColor(palette: StoryArcPalette): Color = when (this) {
    is SourceConnectionState.Connected -> StoryArcColor.Status.success
    is SourceConnectionState.Connecting -> palette.textTertiary
    is SourceConnectionState.Unreachable -> StoryArcColor.Status.offline
    is SourceConnectionState.Unauthorized -> StoryArcColor.Status.danger
}

File: apps/ios/Packages/StoryArcKit/Sources/LibraryFeature/SourcePresentation.swift
internal import SwiftUI

internal import DesignSystem
internal import StoryArcCore

/// Presentation for the domain's source types. Kept out of `StoryArcCore` so the
/// domain stays free of SwiftUI — the same split Android keeps between its
/// `core:model` and `core:designsystem` modules.
extension SourceKind {
    /// SF Symbols only. DESIGN.md §8: no custom icon set, except where the
    /// platform genuinely offers nothing — `opdsCatalog` is the closest fit
    /// available and is revisited if a better symbol ships.
    var symbolName: String {
        switch self {
        case .localFolder: "folder"
        case .networkShare: "externaldrive.connected.to.line.below"
        case .opdsCatalog: "dot.radiowaves.up.forward"
        case .kavitaServer: "server.rack"
        }
    }

    var titleKey: LocalizedStringKey {
        switch self {
        case .localFolder: "source.kind.localFolder.title"
        case .networkShare: "source.kind.networkShare.title"
        case .opdsCatalog: "source.kind.opdsCatalog.title"
        case .kavitaServer: "source.kind.kavitaServer.title"
        }
    }

    var explanationKey: LocalizedStringKey {
        switch self {
        case .localFolder: "source.kind.localFolder.explanation"
        case .networkShare: "source.kind.networkShare.explanation"
        case .opdsCatalog: "source.kind.opdsCatalog.explanation"
        case .kavitaServer: "source.kind.kavitaServer.explanation"
        }
    }
}

extension SourceConnectionState {
    var statusKey: LocalizedStringKey {
        switch self {
        case .connected: "source.state.connected"
        case .connecting: "source.state.connecting"
        case .unreachable: "source.state.unreachable"
        case .unauthorized: "source.state.unauthorized"
        }
    }

    /// Only `unauthorized` is red. An unreachable source is grey, because
    /// `sources` treats offline as a normal state rather than a failure.
    func indicatorColor(_ palette: Palette) -> Color {
        switch self {
        case .connected: StoryArcColor.Status.success
        case .connecting: palette.textTertiary
        case .unreachable: StoryArcColor.Status.offline
        case .unauthorized: StoryArcColor.Status.danger
        }
    }
}

/// How the browsing enums are named on screen.
///
/// The enums themselves live in the domain and carry no strings: `StoryArcCore`
/// has no bundle and no business holding UI copy. Naming them is presentation,
/// so it lives beside the other presentation.
extension LibrarySort {
    var titleKey: LocalizedStringKey {
        switch self {
        case .title: "library.sort.title"
        case .series: "library.sort.series"
        case .lastRead: "library.sort.lastRead"
        case .progress: "library.sort.progress"
        case .year: "library.sort.year"
        }
    }
}

extension ReadState {
    var titleKey: LocalizedStringKey {
        switch self {
        case .unread: "library.readState.unread"
        case .inProgress: "library.readState.inProgress"
        case .finished: "library.readState.finished"
        }
    }
}

extension MatchKind {
    /// The heading over a group of search results.
    ///
    /// Plural, because a heading names a set: "Series" over one result still reads
    /// correctly, where "Serie" over four would not.
    var titleKey: LocalizedStringKey {
        switch self {
        case .series: "library.match.series"
        case .publication: "library.match.publication"
        case .person: "library.match.person"
        case .tag: "library.match.tag"
        }
    }
}
